### ISSUE TITLE
Assign to undef detected during run but not compile 

### ISSUE THREAD

**Issue by philiprbrenan** [Perl/perl5#18779]
The following error message is issued during execution - might it be possible to detect it at compile time so that it can be fixed as soon as possible?  
```
undef = my $a; # Modification of a read-only value attempted
```

**Comment by richardleach** [comment]
I'm not sure what the desirable behavior should be or whether this request should be implemented, but did at least have a quick go at it. A compile time error can be generated with the following:

```
diff --git a/op.c b/op.c
index 594d4ee9c3..61f7eaad75 100644
--- a/op.c
+++ b/op.c
@@ -4291,6 +4291,8 @@ Perl_op_lvalue_flags(pTHX_ OP *o, I32 type, U32 flags)
 
     switch (o->op_type) {
     case OP_UNDEF:
+        if (type == OP_SASSIGN)
+            goto nomod;
        PL_modcount++;
        goto do_next;
```
That produces:
```
# ./miniperl -c -e 'undef = my $foo;'
Can't modify undef operator in scalar assignment at -e line 1, near "$foo;"
-e had compilation errors.
```
I'm running the test suite now to discover what breaks.

**Comment by richardleach** [comment]
> I'm running the test suite now to discover what breaks.

The answer is: only _t/run/fresh_perl.t_, which checks for the runtime error via `BEGIN { undef = 0 }`. That test could be changed as part of any PR for the above. 

Happy to submit a PR if some of the core team could weigh in on whether or not that PR is wanted.

**Comment by hvds** [comment]
Sounds like a good idea to me. Obviously we must not break the list form, but the diff makes clear it won't do that (and we have ample tests for it).

The risk is that people's code could currently be catching this with eval(), maybe without being aware of the fact; it's entirely possible there will be some fallout.

**Comment by Grinnz** [comment]
I think it is a good idea. I find it highly unlikely anyone would rely on it being a runtime rather than compile time error outside of the case found in the test suite. It's rather easy to [grep for on CPAN](https://grep.cpanauthors.org/search?q=%22undef+%3D%22), just have to filter out the false positives, which are mainly: variables named undef, undef followed by pod declarations, or docs.

**Comment by richardleach** [comment]
Thanks @hvds & @Grinnz, I'll self-assign.

**Comment by richardleach** [comment]
A couple of CPAN authors, notably the author of [DTA::CAB::Client::CGIWrapper](https://metacpan.org/pod/release/MOOCOW/DTA-CAB-1.115/CAB/Client/CGIWrapper.pm), seem to use assigning to `undef` in comments and pod for functions that don't have a specific `return` statement. For example:

```
## Methods: top-level
 
undef = $wr->run(\*STDIN);
undef = $wr->finish();
```
Another example is in [Text::MicroMason::Cache::Null](https://metacpan.org/pod/Text::MicroMason::Cache::Null):
```
undef = $cache->get( $key );
```
But are there any circumstances in which that _wouldn't_ trigger the runtime error?

**Comment by Grinnz** [comment]
Right, that's the documentation use I mentioned. It's a convention for notating the return value and as you say wouldn't ever work.

**Comment by richardleach** [comment]
Cool. I appreciate that it's a convention but can imagine unfamiliar users copying verbatim and then wondering why their code failed. :/

**Comment by Grinnz** [comment]
Which we can't really help, but at least a compile time error would make it clear sooner.

**Comment by xenu** [comment]
>But are there any circumstances in which that wouldn't trigger the runtime error?

In perls <= 5.20 it was possible to do this:
```perl
Internals::SvREADONLY(${\undef}, 0);
undef = 7;
print undef; # prints "7"
```

This trick doesn't work anymore because `PL_sv_no` has the `SVf_PROTECT` flag set (see a623f8939cbcaa58a069807591675c0ebcd4135b).

**Comment by richardleach** [comment]
> This trick doesn't work anymore because `PL_sv_no` has the `SVf_PROTECT` flag set (see [a623f89](https://github.com/Perl/perl5/commit/a623f8939cbcaa58a069807591675c0ebcd4135b)).

Aha. I came across that flag the other day, interesting to see it in action.

**Comment by hvds** [comment]
#18785 has now been merged as 2583ef601d; I'll leave this ticket open for a day or two pending smoke results.

**Comment by hvds** [comment]
Smokes look clean; there may be BBC fallout, but we'll get separate tickets if so, so closing this one.
